src/Microsoft.NetCore.Analyzers/Core/Tasks/DoNotCreateTaskCompletionSourceWithWrongArguments.cs

```diff
@@ -0,0 +1,89 @@
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessage,
+                                                                             DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: null,
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // Ensure necessary types are available
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskCompletionSource, out var tcsType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemObject, out var objectType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskContinuationOptions, out var taskContinutationOptionsType))
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    // Ensure that you're using the TaskCompletionSource constructor with the object state property else exit
+                    var invocation = (IObjectCreationOperation)operationContext.Operation;
+                    if (!invocation.Type.OriginalDefinition.Equals(tcsType) ||
+                        invocation.Arguments.Length == 0 ||
+                        !invocation.Constructor.Parameters[0].Type.Equals(objectType))
+                    {
+                        return;
+                    }
+
+                    // Check to see if argument is a conversion operator
+                    if (!(invocation.Arguments[0] is IArgumentOperation arg) ||
```

Comment:
Redundant check given the arguments are always non-null `IArgumentOperation`. Can be rewritten to:

```csharp
if (!(invocation.Arguments[0].Value is IConversionOperation conversionOperation))
{
    return;
}
```